src/engine/enginemaster.cpp

```diff
@@ -443,48 +488,184 @@ void EngineMaster::process(const int iBufferSize) {
+            // to be able to process the same effects on different buffers
+            // within the same callback. For consistency between the MicMonitorModes,
+            // process master effects here before mixing in talkover.
+            applyMasterEffects(iBufferSize, iSampleRate);
+
+            // Mix talkover with master
+            if (m_pNumMicsConfigured->get() > 0) {
+                SampleUtil::copy2WithGain(m_pMaster,
+                    m_pMaster, 1.0,
+                    m_pTalkover, 1.0,
                     iBufferSize);
-        }
+            }
 
-        // Process master channel effects
-        if (m_pEngineEffectsManager) {
-            GroupFeatureState masterFeatures;
-            // Well, this is delayed by one buffer (it's dependent on the
-            // output). Oh well.
-            if (m_pVumeter != NULL) {
-                m_pVumeter->collectFeatures(&masterFeatures);
+            // Copy master mix (with talkover mixed in) to booth output with booth gain
+            if (boothEnabled) {
+                CSAMPLE boothGain = m_pBoothGain->get();
+                if (m_bRampingGain) {
+                    SampleUtil::copy1WithRampingGain(m_pBooth, m_pMaster,
+                        m_boothGainOld, boothGain, iBufferSize);
+                } else {
+                    SampleUtil::copy1WithGain(m_pBooth, m_pMaster,
+                        boothGain, iBufferSize);
+                }
+                m_boothGainOld = boothGain;
+            }
+
+            // Apply master gain
+            // TODO(Be): make this not affect the headphones. Refer to
+            // https://bugs.launchpad.net/mixxx/+bug/1458213
+            CSAMPLE master_gain = m_pMasterGain->get();
+            if (m_bRampingGain) {
+                SampleUtil::applyRampingGain(m_pMaster, m_masterGainOld, master_gain,
+                                             iBufferSize);
+            } else {
+                SampleUtil::applyGain(m_pMaster, master_gain, iBufferSize);
+            }
+            m_masterGainOld = master_gain;
+
+            // Record/broadcast signal is the same as the master output
+            m_ppSidechainOutput = &m_pMaster;
+        } else if (configuredMicMonitorMode == MicMonitorMode::DIRECT_MONITOR) {
+            // Skip mixing talkover with the master and booth outputs
+            // if using direct monitoring because it is being mixed in hardware
+            // without the latency of sending the signal into Mixxx for processing.
+            // However, include the talkover mix in the record/broadcast signal.
+
+            // Copy master mix to booth output with booth gain
+            if (boothEnabled) {
+                CSAMPLE boothGain = m_pBoothGain->get();
+                if (m_bRampingGain) {
+                    SampleUtil::copy1WithRampingGain(m_pBooth, m_pMaster,
+                        m_boothGainOld, boothGain, iBufferSize);
```

Comment:
AFAIK there are only two cases in which direct monitoring are not available: USB microphones and onboard sound cards. There has been a warning on the wiki to avoid USB microphones for a little while, but I will elaborate on this for the 2.1 manual. Then that use case should not be much of an issue (of course there will always be users who buy hardware first then check if it is supported or buy hardware then decide to use Mixxx later). As for onboard sound cards, users are probably not in a loud environment with such a setup and using the direct monitoring mode to disable software monitoring and relying just on monitoring through the air/bone conduction should be good enough.